druid/src/widget/stack.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that arranges its children on top of one another.
+
+use crate::kurbo::{Point, Rect, Size};
+use crate::{
+    BoxConstraints, BoxedWidget, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx,
+    PaintCtx, UpdateCtx, Widget, WidgetPod,
+};
+
+/// A container that lays out its children along the z-axis, first child at bottom, last child on top.
+#[derive(Default)]
+pub struct Stack<T> {
+    children: Vec<BoxedWidget<T>>,
+}
+
+// struct ChildWidget {
+//     widget: BoxedWidget<T>,
+//     interactive: bool,
+// }
+
+impl<T: Data> Stack<T> {
+    /// Create a new stack layout.
+    ///
+    /// The child widgets are laid out on top of one another, from bottom to top.
+    pub fn new() -> Self {
+        Stack {
+            children: Vec::new(),
+        }
+    }
+
+    /// Builder-style variant of `add_child`.
+    ///
+    /// Convenient for assembling a group of widgets in a single expression.
+    pub fn with_child(mut self, child: impl Widget<T> + 'static) -> Self {
+        self.add_child(child);
+        self
+    }
+
+    /// Add a child widget.
+    ///
+    /// See also `with_child`.
+    pub fn add_child(&mut self, child: impl Widget<T> + 'static) {
+        self.children.push(WidgetPod::new(child).boxed());
+    }
+}
+
+impl<T: Data> Widget<T> for Stack<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        for child in &mut self.children.iter_mut().rev() {
```

Comment:
this looks generally good, but one thing I would want is to treat mouse events specially; they should only be delivered to the topmost widget that they intersect with; this should be as easy as going (for the different mouse events):

```rust
if let Some(active_child) = self.children.iter_mut().rev().find(|child| child.layout_rect().contains(mouse_pos) {
    active_child.event(..);
}
```

And then we can also get rid of the `is_handled` check.